File: scripts/day5-part2-optimized.py
import time

def solve_part2(input_data):
    parts = input_data.strip().split('\n\n')
    
    ranges = []
    for line in parts[0].split('\n'):
        start, end = map(int, line.split('-'))
        ranges.append((start, end))
    
    # Sort and merge overlapping ranges
    ranges.sort(key=lambda x: x[0])
    
    total_count = 0
    current_start = ranges[0][0]
    current_end = ranges[0][1]
    
    for i in range(1, len(ranges)):
        start, end = ranges[i]
        if start <= current_end + 1:
            current_end = max(current_end, end)
        else:
            total_count += current_end - current_start + 1
            current_start = start
            current_end = end
    
    total_count += current_end - current_start + 1
    
    return total_count

# Read input
with open("../inputs/day-5-input.txt", "r") as f:
    input_data = f.read()

# Benchmark
iterations = 1
start_time = time.perf_counter()

for _ in range(iterations):
    result = solve_part2(input_data)

end_time = time.perf_counter()
total_time = end_time - start_time
avg_time_us = (total_time / iterations) * 1_000_000

print(f"Python (Day 5 Part 2 - Optimized)")
print(f"Answer: {result}")
print(f"Average time: {avg_time_us:.3f} microseconds ({iterations} iterations)")
print(f"Total time: {total_time * 1000:.2f} ms")

File: scripts/day5-part2-optimized.js
const fs = require("fs");

function solvePart2(input) {
  const ranges = input[0].split("\n").map((line) => {
    const [start, end] = line.split("-").map(Number);
    return { start, end };
  });

  // Sort and merge overlapping ranges
  ranges.sort((a, b) => a.start - b.start);

  let totalCount = 0;
  let currentStart = ranges[0].start;
  let currentEnd = ranges[0].end;

  for (let i = 1; i < ranges.length; i++) {
    if (ranges[i].start <= currentEnd + 1) {
      currentEnd = Math.max(currentEnd, ranges[i].end);
    } else {
      totalCount += currentEnd - currentStart + 1;
      currentStart = ranges[i].start;
      currentEnd = ranges[i].end;
    }
  }

  totalCount += currentEnd - currentStart + 1;

  return totalCount;
}

const input = fs
  .readFileSync("../inputs/day-5-input.txt", "utf-8")
  .trim()
  .split(/\r?\n\r?\n/);

const iterations = 1;
const startTime = process.hrtime.bigint();

let result;
for (let i = 0; i < iterations; i++) {
  result = solvePart2(input);
}

const endTime = process.hrtime.bigint();
const totalNanos = Number(endTime - startTime);
const avgMicros = (totalNanos / iterations / 1000).toFixed(3);

console.log(`JavaScript (Day 5 Part 2 - Optimized)`);
console.log(`Answer: ${result}`);
console.log(`Average time: ${avgMicros} microseconds (${iterations} iterations)`);
console.log(`Total time: ${(totalNanos / 1000000).toFixed(2)} ms`);
